pinecone/index_connection_test.go

```diff
@@ -0,0 +1,161 @@
+package pinecone
+
+import (
+	"context"
+	"github.com/google/uuid"
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/suite"
+	"os"
+	"testing"
+)
+
+type IndexConnectionTests struct {
+	suite.Suite
+	host      string
+	apiKey    string
+	idxConn   *IndexConnection
+	namespace string
+	vectorIds []string
+}
+
+// Runs the test suite with `go test`
+func TestIndexConnection(t *testing.T) {
+	apiKey := os.Getenv("API_KEY")
+
+	podTestSuite := new(IndexConnectionTests)
+	podTestSuite.host = os.Getenv("POD_INDEX_HOST")
+	podTestSuite.apiKey = apiKey
+
+	serverlessTestSuite := new(IndexConnectionTests)
+	serverlessTestSuite.host = os.Getenv("SERVERLESS_INDEX_HOST")
+	serverlessTestSuite.apiKey = apiKey
+
+	suite.Run(t, podTestSuite)
+	suite.Run(t, serverlessTestSuite)
+}
+
+func (ts *IndexConnectionTests) SetupSuite() {
+	assert.NotEmptyf(ts.T(), ts.host, "HOST env variable not set")
+	assert.NotEmptyf(ts.T(), ts.apiKey, "API_KEY env variable not set")
+
+	idxConn, err := newIndexConnection(ts.apiKey, ts.host)
+	assert.NoError(ts.T(), err)
+	ts.idxConn = idxConn
+
+	namespace, err := uuid.NewV7()
+	assert.NoError(ts.T(), err)
+	ts.namespace = namespace.String()
+
+	ts.loadData()
+}
+
+func (ts *IndexConnectionTests) TearDownSuite() {
+	ts.truncateData()
+
+	err := ts.idxConn.Close()
+	assert.NoError(ts.T(), err)
+}
+
+func (ts *IndexConnectionTests) TestFetchVectors() {
+	req := &FetchVectorsRequest{
+		Ids:       ts.vectorIds,
+		Namespace: ts.namespace,
```

Comment:
My golang is rusty, but is there a way to make Namespace optional in these various request objects? On the one hand I like pushing people to be explicit, but on the other hand historically our other SDKs have treated namespace as optional and automatically applied the default `""` when nothing is passed.

In Typescript we tried pulling the namespace up one level so it's more of a config property baked into the thing analgous to your `IndexConnection`. The advantage of that is the user only has to specify it once and doesn't have to remember to pass the correct namespace on every operation. The thinking was that seeing this param repeated across every method suggests it's really part of the connection config, and having to repeat it on every call feels unnecessarily repetitious and error prone. Food for thought.

Reply:
So, as-is these are optional. Golang sets any omitted value to the "zero value" of that type. In `string` type's case, that's `""`.

I do like the idea of capturing an optional `Namespace` in the `IndexConnection` struct and removing it from the individual calls. There aren't cases where folks would be changing the namespace per call, right?

Reply:
> There aren't cases where folks would be changing the namespace per call, right?

It's possible they would want to work with multiple namespaces, but based on what I've seen in our docs and examples it's not common. If somebody wanted to work with multiple namespaces they could create two client instances and it wouldn't be a huge deal, certainly less error prone than having to remember to pass a namespace in every method call. Nobody has complained about this choice in the typescript client AFAIK. As an organizational concept, it's kind of a hangover from pre-serverless days when people needed access to a lighter weight form of multitenancy to save the cost of spinning up a completely separate (expensive) pod index.

Reply:
a84740c

This cleaned up IndexConnection quite a bit! Note, `IndexConnection.Namespace` is exported and can be changed by the user. The new namespace would then be used in subsequent calls.